esrally/storage/_aws.py

```diff
@@ -0,0 +1,162 @@
+# Licensed to Elasticsearch B.V. under one or more contributor
+# license agreements. See the NOTICE file distributed with
+# this work for additional information regarding copyright
+# ownership. Elasticsearch B.V. licenses this file to you under
+# the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import logging
+import os
+import urllib.parse
+from collections.abc import MutableMapping
+from datetime import datetime
+from typing import Any, NamedTuple, TypeVar
+
+import boto3
+import requests
+
+from esrally.storage._adapter import Head, Readable, Writable
+from esrally.storage._http import CHUNK_SIZE, HTTPAdapter
+from esrally.types import Config
+
+LOG = logging.getLogger(__name__)
+
+AWS_PROFILE: str | None = None
+
+A = TypeVar("A", "S3Adapter", "S3Adapter")
+
+
+class S3Adapter(HTTPAdapter):
+    """Adapter class for s3:// scheme protocol"""
+
+    @classmethod
+    def from_config(cls: type[A], cfg: Config, **kwargs: dict[str, Any]) -> A:
+        assert issubclass(cls, S3Adapter)
```

Comment:
Is this necessary? I'm not sure how to call this method without it being a subclass. I would love to have an example of how to accomplish that.